### PR TITLE
Move indices field from RankEvalSpec to RankEvalRequest

### PR DESCRIPTION
**Pull request by cbuescher**
Currently we store the indices specified in the request URL together with all
the other ranking evaluation specification in RankEvalSpec. This is not ideal
since e.g. the indices are not rendered to xContent and so cannot be parsed
back. Instead we should keep them in RankEvalRequest, like we already also do
for other kinds of requests (like SearchRequest).

### CODE REVIEW

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (javanna) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: `in.readStringArray()`?

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (javanna) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: `out.writeStringArray()`?

File: modules/rank-eval/src/test/resources/rest-api-spec/test/rank_eval/10_basic.yml
Comment (javanna) [inline modules/rank-eval/src/test/resources/rest-api-spec/test/rank_eval/10_basic.yml]: is this correct? why skip rest tests? aren't we only changing where indices are set internally?

**Review by javanna — COMMENTED** [review]
I had not seen this, I left a couple of comments, LGTM though. Wasn't there also a TODO somewhere in the high-level client code that can be resolved once we merge this in?

**Comment by cbuescher** [comment]
@javanna thanks for looking into this, I will update the PR shortly

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (cbuescher) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: I changed this.

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (cbuescher) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: I changed this.

File: modules/rank-eval/src/test/resources/rest-api-spec/test/rank_eval/10_basic.yml
Comment (cbuescher) [reply modules/rank-eval/src/test/resources/rest-api-spec/test/rank_eval/10_basic.yml]: I changed this previously because I changed the place where the indices strings for serialized/deserialized. However, I now think this isn't necessary since even though the location where the string array is writen changes, the overall ordering on the wire shouldn't change, thus there should be nothing breaking here. I removed the version skip changes in the last update.

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalSpec.java
Comment (cbuescher) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalSpec.java]: This should me handled by out.writeStringArray(indices) in the request now, and shouldn't change the wire protocol so no version-dependent logic needed here I think.

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalSpec.java
Comment (cbuescher) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalSpec.java]: This should me handled by in.readStringArray() in the request now, and shouldn't change the wire protocol so no version-dependent logic needed here I think.

**Comment by cbuescher** [comment]
@javanna thanks for the review, I pushed an update. Can you take a look again and also check if my above reasoning about not needing a bwc check on the wire makes sense?

**Review by javanna — CHANGES_REQUESTED** [review]
left a small comment

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalSpec.java
Comment (cbuescher) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalSpec.java]: Good point, I will add a version check here.

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (cbuescher) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: I wanted to protect the indices from accidental modifications, but since there is a setter for them, they are mutable anyway, so I changed this like you suggested.

**Comment by cbuescher** [comment]
@javanna I added version checks and made a few more changes in my last updates, can you take another look?

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (javanna) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: add to the comment readVint vs readInt just to clarify something that may not be clear at first?

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (javanna) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: here too

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (javanna) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: on the version conditionals, you may need to first use 7_0_0_alpha1 on master before you backport to 6.x, then you can update the version on master too.

**Review by javanna — APPROVED** [review]

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (cbuescher) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: I don't know if this clarifies things a lot since it requires looking into what readStringArray does? At this point I wonder if I should remove the comment completely, the code should explain this by itself and I think we can remove it on master after 6.3 is release (we only try to be bwc with the latest minor release before a major, no?)

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (cbuescher) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: Given that 6.3 isn't release yet, bwc tests should only run vs. 6.2 if this is merged shortly, or am I missung sth?

Comment (javanna) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: don't we run bwc tests between 6.x and previous versions?

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (javanna) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: I think the comment may be useful, to clarify why we have this conditional as it looks weird at first. Of course if one looks deeper will also find out why we have it. Up to you

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java
Comment (cbuescher) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalRequest.java]: np, I just thought about it when you mentioned it.

**Comment by cbuescher** [comment]
@javanna thanks for the review, I changed the comments slightly, will wait for this CI build before merging

**Comment by cbuescher** [comment]
@elasticmachine test this again please

**Pull request merged**
